src/witness_generator/machines/fixed_lookup_machine.rs

```diff
@@ -14,9 +14,73 @@ use crate::witness_generator::{
     EvalResult, FixedData,
 };
 
+/// Indices for COLUMNS. Stores (V, rows) iff every row in rows satisfied COLUMNS[row] == V
+#[derive(Default)]
+
+pub struct IndexedColumns {
+    indices: HashMap<Vec<String>, BTreeMap<Vec<AbstractNumberType>, Vec<DegreeType>>>,
+}
+
+impl IndexedColumns {
+    /// get the rows at which the assignment is satisfied
+    /// Warning: in particular, an empty assignment will return all rows
+    fn get_matches(
+        &mut self,
+        fixed_data: &FixedData,
+        mut assignment: Vec<(String, AbstractNumberType)>,
+    ) -> Option<&Vec<DegreeType>> {
```

Comment:
This needs to be an option because we don't have a reference to an empty vector, right?

Reply:
Right.. Now the caller uses `flatten`. I guess we could also have a static empty vector we can have a reference to...